File: src/main/services/ModUnInstaller.ts
import fs from 'fs'
import path from 'path'
import type { ModManifestDto } from '../../shared/types'
import {
  bepinexPatchersDir,
  BEPINEX_CONFIG_FOLDER,
  GMP_METADATA_FOLDER,
  TEMP_FOLDER
} from '../constants'
import { isInside } from './PathGuard'

function removeSafe(target: string, root: string): void {
  if (fs.existsSync(target)) {
    if (!isInside(root, target)) {
      throw new Error(`refusing to modify path outside root: ${target}`)
    }
    fs.rmSync(target, { recursive: true, force: true })
  }
}

function removeFileIfExists(target: string, root: string): void {
  if (fs.existsSync(target)) {
    if (!isInside(root, target)) {
      throw new Error(`refusing to modify path outside root: ${target}`)
    }
    if (fs.statSync(target).isFile()) fs.rmSync(target, { force: true })
  }
}


function dllStem(nameOrPath: string): string {
  const base = path.basename(nameOrPath)
  let s = base
  let lc = s.toLowerCase()
  if (lc.endsWith('.disabled')) s = s.slice(0, -9)
  else if (lc.endsWith('.disable')) s = s.slice(0, -8)
  lc = s.toLowerCase()
  const backup = /^(.*)\.dll\.\d+$/i.exec(s)
  if (backup) return backup[1]
  if (lc.endsWith('.dll')) return s.slice(0, -4)
  return s
}


function cfgFiles(gameRoot: string, names: string[]): string[] {
  const configDir = path.join(gameRoot, 'BepInEx', BEPINEX_CONFIG_FOLDER)
  const out: string[] = []
  for (const n of names) {
    const stem = dllStem(n)
    if (stem && !out.includes(`${stem}.cfg`)) out.push(path.join(configDir, `${stem}.cfg`))
  }
  return out
}

interface Backup {
  src: string
  rel: string
}

function backupTargets(
  gameRoot: string,
  backupRoot: string,
  targets: string[]
): Backup[] {
  const backedUp: Backup[] = []
  for (const src of targets) {
    if (fs.existsSync(src)) {
      const rel = path.relative(gameRoot, src)
      const dst = path.join(backupRoot, rel)
      fs.cpSync(src, dst, { recursive: true, force: true })
      backedUp.push({ src, rel })
    }
  }
  return backedUp
}

function restoreBackups(gameRoot: string, backupRoot: string, backedUp: Backup[]): void {
  for (const b of backedUp.reverse()) {
    try {
      const orig = path.join(gameRoot, b.rel)
      if (fs.existsSync(b.src)) fs.rmSync(b.src, { recursive: true, force: true })
      if (fs.existsSync(orig)) fs.rmSync(orig, { recursive: true, force: true })
      fs.mkdirSync(path.dirname(orig), { recursive: true })
      fs.cpSync(path.join(backupRoot, b.rel), orig, { recursive: true, force: true })
    } catch {
      
    }
  }
}






export function deleteMod(gameRoot: string, installDir: string, manifest: ModManifestDto): void {
  const backupRoot = path.join(gameRoot, GMP_METADATA_FOLDER, TEMP_FOLDER, `backup_${Date.now()}`)
  fs.mkdirSync(backupRoot, { recursive: true })

  const pluginNames = [...manifest.plugins]
  const patcherNames = [...manifest.patchers]
  const patchersDir = bepinexPatchersDir(gameRoot)

  let backedUp: Backup[] = []
  try {
    
    const targets: string[] = [installDir]
    for (const a of manifest.assets) {
      if (a.destination) targets.push(path.resolve(gameRoot, a.destination))
    }
    for (const p of patcherNames) {
      targets.push(path.join(patchersDir, path.basename(p)))
    }
    targets.push(...cfgFiles(gameRoot, [...pluginNames, ...patcherNames]))
    backedUp = backupTargets(gameRoot, backupRoot, [...new Set(targets)])

    
    removeSafe(installDir, gameRoot)

    
    for (const p of patcherNames) {
      const abs = path.join(patchersDir, path.basename(p))
      removeSafe(abs, gameRoot)
      removeFileIfExists(abs.replace(/\.dll$/i, '.disabled'), gameRoot)
    }

    
    for (const a of manifest.assets) {
      if (!a.destination) continue
      const dest = path.resolve(gameRoot, a.destination)
      if (isInside(gameRoot, dest)) removeSafe(dest, gameRoot)
    }

    
    for (const c of cfgFiles(gameRoot, [...pluginNames, ...patcherNames])) {
      removeFileIfExists(c, gameRoot)
    }

    
    fs.rmSync(backupRoot, { recursive: true, force: true })
  } catch (e) {
    restoreBackups(gameRoot, backupRoot, backedUp)
    throw e
  }
}






export function deleteLegacyPlugin(
  gameRoot: string,
  installDir: string,
  pluginFiles: string[]
): void {
  const pluginsDir = path.join(gameRoot, 'BepInEx', 'plugins')
  const backupRoot = path.join(gameRoot, GMP_METADATA_FOLDER, TEMP_FOLDER, `backup_${Date.now()}`)
  fs.mkdirSync(backupRoot, { recursive: true })

  const stems = pluginFiles.map(dllStem)
  const lookups: string[] = []

  
  for (const pf of pluginFiles) {
    const abs = path.resolve(installDir, pf)
    if (isInside(installDir, abs) && isInside(gameRoot, abs)) lookups.push(abs)
  }
  
  const scanRoot =
    installDir !== pluginsDir && isInside(pluginsDir, installDir) && fs.existsSync(installDir)
      ? installDir
      : pluginsDir
  if (fs.existsSync(scanRoot)) {
    for (const e of fs.readdirSync(scanRoot, { withFileTypes: true })) {
      const stem = dllStem(e.name)
      if (stems.includes(stem)) {
        const abs = path.resolve(scanRoot, e.name)
        if (isInside(gameRoot, abs) && isInside(pluginsDir, abs)) lookups.push(abs)
      }
    }
  }

  const targets = [...new Set(lookups)]
  targets.push(...cfgFiles(gameRoot, stems))

  let backedUp: Backup[] = []
  try {
    backedUp = backupTargets(gameRoot, backupRoot, targets)
    for (const t of targets) {
      if (!fs.existsSync(t)) continue
      if (!isInside(gameRoot, t)) throw new Error(`refusing to modify path outside root: ${t}`)
      if (fs.statSync(t).isFile()) fs.rmSync(t, { force: true })
      else removeSafe(t, gameRoot)
    }
    
    if (
      installDir !== pluginsDir &&
      isInside(pluginsDir, installDir) &&
      fs.existsSync(installDir) &&
      fs.readdirSync(installDir).length === 0
    ) {
      fs.rmdirSync(installDir)
    }
    fs.rmSync(backupRoot, { recursive: true, force: true })
  } catch (e) {
    restoreBackups(gameRoot, backupRoot, backedUp)
    throw e
  }
}

File: RBBPMM.Wpf/RBBPMM.Core/ModUnInstaller.cs
namespace RBBPMM.Core;

/// <summary>Safe removal of installed mods with a backup-and-restore safety net.</summary>
public static class ModUnInstaller
{
    private const string PluginExt = ".dll";

    private static void RemoveSafe(string target, string root)
    {
        if (Directory.Exists(target))
        {
            if (!PathGuard.IsInside(root, target))
                throw new InvalidOperationException($"refusing to modify path outside root: {target}");
            Directory.Delete(target, recursive: true);
        }
    }

    private static void RemoveFileIfExists(string target, string root)
    {
        if (File.Exists(target))
        {
            if (!PathGuard.IsInside(root, target))
                throw new InvalidOperationException($"refusing to modify path outside root: {target}");
            File.Delete(target);
        }
    }

    private static List<string> CfgFiles(string gameRoot, IEnumerable<string> names)
    {
        var configDir = Path.Combine(gameRoot, "BepInEx", Constants.BepInExConfigFolder);
        var outList = new List<string>();
        foreach (var n in names)
        {
            var stem = ModActivator.DllStem(n);
            if (!string.IsNullOrEmpty(stem) && !outList.Contains($"{stem}.cfg"))
                outList.Add(Path.Combine(configDir, $"{stem}.cfg"));
        }
        return outList;
    }

    private sealed record Backup(string Src, string Rel);

    private static List<Backup> BackupTargets(string gameRoot, string backupRoot, List<string> targets)
    {
        var backedUp = new List<Backup>();
        foreach (var src in targets)
        {
            if (!File.Exists(src) && !Directory.Exists(src)) continue;
            var rel = Path.GetRelativePath(gameRoot, src);
            var dst = Path.Combine(backupRoot, rel);
            Directory.CreateDirectory(Path.GetDirectoryName(dst)!);
            if (Directory.Exists(src))
                CopyDirectory(src, dst);
            else
                File.Copy(src, dst, overwrite: true);
            backedUp.Add(new Backup(src, rel));
        }
        return backedUp;
    }

    private static void RestoreBackups(string gameRoot, string backupRoot, List<Backup> backedUp)
    {
        foreach (var b in Enumerable.Reverse(backedUp))
        {
            try
            {
                var orig = Path.Combine(gameRoot, b.Rel);
                if (File.Exists(b.Src)) File.Delete(b.Src);
                if (Directory.Exists(b.Src)) Directory.Delete(b.Src, recursive: true);
                var src = Path.Combine(backupRoot, b.Rel);
                Directory.CreateDirectory(Path.GetDirectoryName(orig)!);
                if (Directory.Exists(src))
                    CopyDirectory(src, orig);
                else if (File.Exists(src))
                    File.Copy(src, orig, overwrite: true);
            }
            catch { /* best effort */ }
        }
    }

    public static void DeleteMod(string gameRoot, string installDir, ModManifest manifest)
    {
        var backupRoot = Path.Combine(gameRoot, Constants.GmpMetadataFolder, Constants.TempFolder,
            $"backup_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}");
        Directory.CreateDirectory(backupRoot);
        var patchersDir = Constants.BepinexPatchersDir(gameRoot);
        List<Backup> backedUp = [];
        try
        {
            var targets = new List<string> { installDir };
            foreach (var a in manifest.Assets)
                if (!string.IsNullOrEmpty(a.Destination))
                    targets.Add(Path.GetFullPath(Path.Combine(gameRoot, a.Destination)));
            foreach (var p in manifest.Patchers)
                targets.Add(Path.Combine(patchersDir, Path.GetFileName(p)));
            targets.AddRange(CfgFiles(gameRoot, manifest.Plugins.Concat(manifest.Patchers)));

            backedUp = BackupTargets(gameRoot, backupRoot, targets.Distinct().ToList());
            RemoveSafe(installDir, gameRoot);
            foreach (var p in manifest.Patchers)
            {
                var abs = Path.Combine(patchersDir, Path.GetFileName(p));
                RemoveSafe(abs, gameRoot);
                RemoveFileIfExists(abs[..^PluginExt.Length] + ".disabled", gameRoot);
            }
            foreach (var a in manifest.Assets)
            {
                if (string.IsNullOrEmpty(a.Destination)) continue;
                var dest = Path.GetFullPath(Path.Combine(gameRoot, a.Destination));
                if (PathGuard.IsInside(gameRoot, dest))
                    RemoveSafe(dest, gameRoot);
            }
            foreach (var c in CfgFiles(gameRoot, manifest.Plugins.Concat(manifest.Patchers)))
                RemoveFileIfExists(c, gameRoot);
            Directory.Delete(backupRoot, recursive: true);
        }
        catch
        {
            RestoreBackups(gameRoot, backupRoot, backedUp);
            throw;
        }
    }

    public static void DeleteLegacyPlugin(string gameRoot, string installDir, List<string> pluginFiles)
    {
        var pluginsDir = Path.Combine(gameRoot, "BepInEx", Constants.PluginsFolder);
        var backupRoot = Path.Combine(gameRoot, Constants.GmpMetadataFolder, Constants.TempFolder,
            $"backup_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}");
        Directory.CreateDirectory(backupRoot);
        var stems = pluginFiles.Select(ModActivator.DllStem).ToList();
        var lookups = new List<string>();
        foreach (var pf in pluginFiles)
        {
            var abs = Path.GetFullPath(Path.Combine(installDir, pf));
            if (PathGuard.IsInside(installDir, abs) && PathGuard.IsInside(gameRoot, abs))
                lookups.Add(abs);
        }
        var scanRoot = (installDir != pluginsDir && PathGuard.IsInside(pluginsDir, installDir) && Directory.Exists(installDir))
            ? installDir : pluginsDir;
        if (Directory.Exists(scanRoot))
        {
            foreach (var e in Directory.GetFiles(scanRoot))
            {
                var stem = ModActivator.DllStem(Path.GetFileName(e));
                if (stems.Contains(stem))
                {
                    var abs = Path.GetFullPath(Path.Combine(scanRoot, Path.GetFileName(e)));
                    if (PathGuard.IsInside(gameRoot, abs) && PathGuard.IsInside(pluginsDir, abs))
                        lookups.Add(abs);
                }
            }
        }
        var targets = lookups.Distinct().ToList();
        targets.AddRange(CfgFiles(gameRoot, stems));
        List<Backup> backedUp = [];
        try
        {
            backedUp = BackupTargets(gameRoot, backupRoot, targets);
            foreach (var t in targets)
            {
                if (!File.Exists(t) && !Directory.Exists(t)) continue;
                if (!PathGuard.IsInside(gameRoot, t))
                    throw new InvalidOperationException($"refusing to modify path outside root: {t}");
                if (File.Exists(t)) File.Delete(t);
                else RemoveSafe(t, gameRoot);
            }
            if (installDir != pluginsDir && PathGuard.IsInside(pluginsDir, installDir) &&
                Directory.Exists(installDir) && Directory.GetFileSystemEntries(installDir).Length == 0)
                Directory.Delete(installDir);
            Directory.Delete(backupRoot, recursive: true);
        }
        catch
        {
            RestoreBackups(gameRoot, backupRoot, backedUp);
            throw;
        }
    }

    private static void CopyDirectory(string src, string dst)
    {
        Directory.CreateDirectory(dst);
        foreach (var entry in Directory.GetFileSystemEntries(src))
        {
            var name = Path.GetFileName(entry);
            var to = Path.Combine(dst, name);
            if (Directory.Exists(entry))
                CopyDirectory(entry, to);
            else
                File.Copy(entry, to, overwrite: true);
        }
    }
}
